src/review/commands/build.ts

```diff
@@ -0,0 +1,128 @@
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Command, flags } from '@oclif/command';
+import * as fg from 'fast-glob';
+import * as fileSize from 'filesize';
+import { promises, statSync } from 'fs';
+import * as Path from 'path';
+import {
+  getPackageJsonDeps,
+  getPackageLockDeps,
+  nodeModulesGlobs,
+} from '../bundle';
+import { compileJsModule } from '../bytecode';
+import { Entry, write } from '../format';
+
+// tslint:disable-next-line: no-default-export
+export default class Build extends Command {
+  static description = 'build a bundle';
+  static examples = [`$ bulkan build`];
+
+  static flags = {
+    help: flags.help({ char: 'h' }),
+    no_compile: flags.boolean({
+      char: 'C',
+      description: "don't compile javascript files",
+    }),
+    outfile: flags.string({
+      char: 'o',
+      description: 'output file for the generated bundle.',
+      default: 'bundle.blkn',
+    }),
+  };
+
+  static args = [{ name: 'cwd', default: './' }];
+
+  async run() {
+    const { args, flags } = this.parse(Build);
+
+    const handlers: { [extension: string]: Handler } = {
+      '.json': loadBuffer,
+      '.js': loadBufferAndCompileCache,
+    };
+    if (flags.no_compile) handlers['.js'] = loadBuffer;
+
+    this.log('Collecting paths');
+    const pkgPath = Path.resolve(Path.join(args.cwd, 'package.json'));
+    const pkgLockPath = Path.resolve(Path.join(args.cwd, 'package-lock.json'));
+    const pkgDeps = getPackageJsonDeps(require(pkgPath));
+    const pkgLockDeps = getPackageLockDeps(require(pkgLockPath));
+    const globs = nodeModulesGlobs([...new Set([...pkgDeps, ...pkgLockDeps])]);
+
+    const paths = await fg(globs);
+    this.log(`Found ${paths.length} paths`);
+
+    this.log(`Preparing buffers`);
+    const entryPs = ([] as Array<Promise<Entry>>)
+      .concat(
+        ...paths.map((path: string) => {
+          const ext = Path.extname(path);
+          if (ext in handlers) return handlers[ext](path);
+          return [];
+        })
+      )
+      .map(e =>
+        e.catch(err => {
+          this.log(`Failed to load ${err.path}: ${err.msg}`);
+          return null;
+        })
+      );
+    const entries = (await Promise.all(entryPs)).filter(e => !!e) as Entry[];
+
+    this.log(`Writing bundle to ${flags.outfile}`);
+    write(flags.outfile as string, entries);
+
+    const size = fileSize(statSync(flags.outfile as string).size);
+    this.log(
+      `Bundled ${entries.length} files into ${flags.outfile} using ${size}`
+    );
+  }
+}
+
+type Handler = (path: string) => Array<Promise<Entry>>;
+
+const loadBuffer: Handler = path => {
```

Comment:
same nit: its not clear why one function returns an array with on item vs the other which returns an array of 2 items. it feels like you're wrapping the expected value with superfluous data structures to fit into the processing code which lives somewhere else rather than keeping these functions focused on what they do.

Reply:
`loadBuffer` and `loadBufferAndCompileCache` are `Handler`s, meaning they return (promised) `Entry`s (of which there can be 0 or more) that correspond to a path. Given that context, and that they're local to this module, I think it makes sense.